File: go/86.go
package main

func partition(head *ListNode, x int) *ListNode {
	beforeDummy := &ListNode{}
	before := beforeDummy
	afterDummy := &ListNode{}
	after := afterDummy

	for head != nil {
		if head.Value < x {
			before.Next = head
			before = before.Next
		} else {
			after.Next = head
			after = after.Next
		}

		head = head.Next
	}

	after.Next = nil
	before.Next = afterDummy.Next

	return beforeDummy.Next
}

File: zig/86.zig
const std = @import("std");

const ListNode = struct {
    val: i32,
    next: ?*ListNode,
};

fn solution(node: ?*ListNode, x: i32) ?*ListNode {
    const beforeDummy = &ListNode{};
    const afterDummy = &ListNode{};

    var before = beforeDummy;
    var after = afterDummy;

    var runner = node;

    while (runner) |n| {
        if (n.val < x) {
            before.next = n;
            before = before.next;
        } else {
            after.next = n;
            after = after.next;
        }

        runner = runner.next;
    }

    after.next = null;
    before.next = afterDummy.next;

    return beforeDummy.next;
}
